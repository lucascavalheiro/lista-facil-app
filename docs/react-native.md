# Getting Started with React Native

## Setting Up Your Machine for App Development

- iOS (You require a Mac to develop iOS Apps) Simply download XCode
- [Android on Mac](https://medium.com/pvtl/react-native-android-development-on-mac-ef7481f65e47#.w2hel88zy)
- Android on Windows - coming soon...

## Install React Native

[React Native - Get Started Guide](https://facebook.github.io/react-native/docs/getting-started.html) 

## Learn React

[React Native Express](http://www.reactnativeexpress.com/) is a great site to get you started, specifically:

- [Get your head around ES6](http://www.reactnativeexpress.com/es6)
- [What is JSX?](http://www.reactnativeexpress.com/jsx)
- [What are Components?](http://www.reactnativeexpress.com/components)
- [React State](http://www.reactnativeexpress.com/data_component_state)
- [Redux](http://www.reactnativeexpress.com/redux)

Once you've got your head around the basics, checkout the [React Native](https://facebook.github.io/react-native/) website, specifically:

- Go through ['The Basics'](https://facebook.github.io/react-native/docs/props.html)
- Gain an understanding of the [components](https://facebook.github.io/react-native/docs/activityindicator.html) React Native provides out of the box
