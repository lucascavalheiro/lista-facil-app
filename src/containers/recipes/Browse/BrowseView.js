/**
 * Receipe Tabs Screen
 *  - Shows tabs, which contain receipe listings
 *
 * React Native Starter App
 * https://github.com/mcnamee/react-native-starter-app
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  View,
  StyleSheet,
  InteractionManager,
} from 'react-native';
import { TabViewAnimated, TabBar } from 'react-native-tab-view';

// Consts and Libs
import { AppColors } from '@theme/';

// Containers
import RecipeListing from '@containers/recipes/Listing/ListingContainer';

// Components
import { Text } from '@ui/';
import Loading from '@components/general/Loading';

/* Styles ==================================================================== */
const styles = StyleSheet.create({
  // Tab Styles
  tabContainer: {
    flex: 1,
  },
  tabbar: {
    backgroundColor: AppColors.brand.primary,
  },
  tabbarIndicator: {
    backgroundColor: '#FFF',
  },
  tabbarText: {
    color: '#FFF',
  },
});

/* Component ==================================================================== */
let loadingTimeout;
class RecipeTabs extends Component {
  static componentName = 'RecipeTabs';

  static propTypes = {
    meals: PropTypes.arrayOf(PropTypes.object).isRequired,
  }

  static defaultProps = {
    meals: [],
  }

  constructor(props) {
    super(props);

    this.state = {
      loading: true,
      visitedRoutes: [],
    };
  }

  /**
    * Wait until any interactions are finished, before setting up tabs
    */
  componentDidMount = () => {
    InteractionManager.runAfterInteractions(() => {
      this.setTabs();
    });
  }

  componentWillUnmount = () => clearTimeout(loadingTimeout);

  /**
    * When meals are ready, populate tabs
    */
  setTabs = () => {
    const routes = [];
    let idx = 0;
    this.props.meals.forEach((meal) => {
      routes.push({
        key: idx.toString(),
        id: meal.id.toString(),
        title: meal.title,
      });

      idx += 1;
    });

    this.setState({
      navigation: {
        index: 0,
        routes,
      },
    }, () => {
      // Hack to prevent error showing
      loadingTimeout = setTimeout(() => {
        this.setState({ loading: false });
      }, 100);
    });
  }

  /**
    * On Change Tab
    */
  handleChangeTab = (index) => {
    this.setState({
      navigation: { ...this.state.navigation, index },
    });
  }

  /**
    * Header Component
    */
  renderHeader = props => (
    <TabBar
      {...props}
      style={styles.tabbar}
      indicatorStyle={styles.tabbarIndicator}
      renderLabel={scene => (
        <Text style={[styles.tabbarText]}>{scene.route.title}</Text>
      )}
    />
  )

  /**
    * Which component to show
    */
  renderScene = ({ route }) => {
    // For performance, only render if it's this route, or I've visited before
    if (
      parseInt(route.key, 0) !== parseInt(this.state.navigation.index, 0) &&
      this.state.visitedRoutes.indexOf(route.key) < 0
    ) {
      return null;
    }

    // And Add this index to visited routes
    if (this.state.visitedRoutes.indexOf(this.state.navigation.index) < 0) {
      this.state.visitedRoutes.push(route.key);
    }

    // Which component should be loaded?
    return (
      <View style={styles.tabContainer}>
        <RecipeListing
          meal={route.id}
        />
      </View>
    );
  }

  render = () => {
    if (this.state.loading || !this.state.navigation) return <Loading />;

    return (
      <TabViewAnimated
        style={[styles.tabContainer]}
        renderScene={this.renderScene}
        renderHeader={this.renderHeader}
        navigationState={this.state.navigation}
        onRequestChangeTab={this.handleChangeTab}
      />
    );
  }
}

/* Export Component ==================================================================== */
export default RecipeTabs;
