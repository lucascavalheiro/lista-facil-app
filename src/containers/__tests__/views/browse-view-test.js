/**
 * Test to check if the component renders correctly
 */
/* global it expect */
import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import BrowseView from '@containers/recipes/Browse/BrowseView';

it('BrowseView renders correctly', () => {
  // getMeals prop expects a promise
  const mockPromise = new Promise((resolve) => {
    resolve();
  });

  const tree = renderer.create(
    <BrowseView getMeals={() => mockPromise} />,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
