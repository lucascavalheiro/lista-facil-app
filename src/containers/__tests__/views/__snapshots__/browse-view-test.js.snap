// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowseView renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#E9EBEE",
        "flex": 1,
        "flexDirection": "column",
        "position": "relative",
      },
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      false,
    ]
  }
>
  <ActivityIndicator
    animating={true}
    color="#AAA"
    hidesWhenStopped={true}
    size="large"
  />
  <View
    style={
      Object {
        "height": 1,
        "left": 0,
        "marginTop": 9,
        "right": 0,
      }
    }
  />
</View>
`;
